### DISCUSSION TITLE
Update edge from target handle, not end of the edge

### DISCUSSION THREAD

**Discussion by ambika-dubey** [xyflow/xyflow#1350] [Q&A]
I believe the current default behavior is to enable edge updates/edits from the end of an existing edge (not from the target handle that the edge connects to). Is there a way either to set some flag so that updates can happen from the target handle, or a straightforward workaround? Thank you!

**Comment by mikeestee** [comment]
> I believe the current default behavior is to enable edge updates/edits from the end of an existing edge (not from the target handle that the edge connects to). Is there a way either to set some flag so that updates can happen from the target handle, or a straightforward workaround? Thank you!

Further context: we're using this behavior during "edge move" operations, where the user switches an edge from one handle to another. The desired behavior is for the user to be able to "pick up" an edge and drag it over to a new handle.

**Reply by mikeestee** [reply]
@moklick  From further examination of the code base it looks like this is not possible to implement without changes to react-flow. We could use advice on the best approach for this.

I'm thinking:
- Start an edge move operation from a handle if the handle is connected. Hide existing edge for duration of the drag, letting ConnectionEdge represent the edge while dragging.
- Start a new edge operation from a handle if the handle is unconnected.

**Reply by ambika-dubey** [reply]
What we can do is use onEdgeUpdateStart and onEdgeUpdateEnd to toggle the isHidden flag on the edge being updated to get the "picking up the edge" behavior described above.
![edgeDrag](https://user-images.githubusercontent.com/9057054/125984985-3ad42807-c3b1-4bd4-9283-8fa06868ba38.gif)

However, any help with starting an edge update from a connected handle would be extremely helpful!
To help illustrate what we're trying to do, as explained above, when clicking and dragging from a connected handle, current behavior is to start a new connection. We'd like to be able to make that an edge update operation:
![behavior](https://user-images.githubusercontent.com/9057054/126018900-d70612d6-2bef-47cb-90c9-4ed7f2478998.gif)

**Reply by moklick** [reply]
When I get it right you would like to disable the default behaviour (create a new edge) and enable an edge update when the user starts to drag on a handle. At the moment this is not possible. The problem that I see with this behaviour is that it does different things. When there is a connected edge I can update it. If there is none I create a new one.

The question is how can we support this functionality? Do you have an idea? Maybe a flag `updateEdgeOnHandleDrag`, `updateEdgeWhenConnected`, . .I can't think of a good name

**Reply by ambika-dubey** [reply]
I think a flag like that (`updateEdgeWhenConnected` or `updateConnectedEdgeFromHandle`) would be ideal, or some combination of:
- an even stricter mode for edge connections, where dragging to create a new node is only possible from source handles, not target handles, and:
- some sort of override for handle drag behavior so that it can function as the end of the edge to start an update operation when it has a connected edge

But, I can 100% see issues with either of these--we're assuming one edge per target handle, but that is something we're enforcing for our use-case, and it's not correct for everyone. Say we implement a flag `updateEdgeWhenConnected` but there are multiple edges connected to a target handle, do we do a batch update for all of the edges? Choose one to update? Update only if there is strictly one edge?

I'm new to function components and hooks, but maybe providing a hook like `startEdgeUpdateOperation` that takes in an edgeId and maybe a position so we know where to draw the end of the in-progress connection might work?

@mikeestee any thoughts here?

**Reply by mikeestee** [reply]
I think something that allows the developer to conditionally change the operation for various handle/edge operations is probably warranted.

Much like there is a connection validation step where the developer can conditionally allow a connection to be made, there could be a callback to ask what the operation should be (move existing, start new, delete old) when the user initiates an action from a handle.

This would let the developer customize a variety of behaviors, including operations for multi-target and multi-source. I would expect that this callback would take a list of edges currently connected so it can make a decision about the behavior. I can imagine it returning an operation and list of edges to operate on.

spitballing with some pseudocode:

```javascript
onHandleAction(node: Node, handle: Handle, edges: Elements[]): {action: EdgeAction, edges: Elements[]} {
  ...
  if( shiftKeyIsDown() )
      return {action: MoveEdges, edges};
  else
     return {action: NewEdge, edges: undefined};
}
```
